spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+another. This means applications need to ensure access to state is
+thread safe. Up to v0.35.x, both the
+[default in-process ABCI client](https://github.com/tendermint/tendermint/blob/v0.35.x/abci/client/local_client.go#L18)
+and the
+[default Go ABCI server](https://github.com/tendermint/tendermint/blob/v0.35.x/abci/server/socket_server.go#L32)
+used global locks to guard the handling of events across all connections, so they were not
+concurrent at all. This meant if your app was written in Go, and compiled in-process with Tendermint
+using the default `NewLocalClient`, or run out-of-process using the default `SocketServer`,
+ABCI messages from all connections were linearizable (received one at a
+time).
+This is no longer the case starting from v0.36.1: the global locks have been removed and it is
+up to the Application to ensure that handling ABCI++ connections is thread safe.
+
+#### FinalizeBlock
+
+When the Consensus algorithm decides on a block, Tendermint uses `FinalizeBlock` to send the
+current block's hash, height, transactions, etc., which the Application uses to transition its state.
+
+The Application should remember the latest height from which it
+has run a successful `Commit` so that it can tell Tendermint where to
+pick up from when it restarts. See information on the Handshake, below.
+
+#### Commit
+
+The Application state should be persisted during `Commit`.
+
+Before invoking `Commit`, Tendermint locks the mempool and flushes the mempool connection. This ensures that
+no new messages
+will be received on the mempool connection during this processing step, providing a good opportunity to safely update all four
+connection states to the latest committed state at once.
+
+When `Commit` completes, Tendermint unlocks the mempool.
+
+WARNING: if the ABCI app logic processing the `Commit` message sends a
+`/broadcast_tx_sync` or `/broadcast_tx_commit` and waits for the response
+before proceeding, it will deadlock. Executing `broadcast_tx` calls
+involves acquiring the mempool lock that Tendermint holds during the `Commit` call.
+If you make the call to the `broadcast_tx` endpoints concurrently,
+that's not a problem, it just can't be part of the sequential logic of the
+`Commit` function.
+
+#### Candidate States
+
+Tendermint calls `PrepareProposal` when it is about to send a proposed block to the network.
+Likewise, Tendermint calls `ProcessProposal` upon reception of a proposed block from the
+network. In both cases, the proposed block's data (last block's commit info, transactions, etc.)
+is disclosed to the Application, in the same conditions as is done in `FinalizeBlock`.
+
+The Application may decide to *immediately* execute the given block (i.e., upon `PrepareProposal`
+or `ProcessProposal`). There are two main reasons why the Application may want to do this:
+
+* In order to be sure that the block does not contain *any* invalid transaction, there is usually
```

Comment:
"usually? is unclear.

Reply:
"usually", because there may be cases of (simple) applications that don't need to fully execute the transactions to detect the invalid ones.
Anyway, I reworded